tensorpack/tfutils/distributions.py

```diff
@@ -0,0 +1,126 @@
+import tensorflow as tf
+from functools import wraps
+import numpy as np
+
+__all__ = ['CategoricalDistribution']
+
+
+def class_scope(method):
+    """Enhance TensorBoard graph visualization by grouping operators.
+
+    If class of method has member "name", then this is used for name-scoping.
+
+    Args:
+        method (TYPE): method from a python class
+
+    Returns:
+        TYPE: Description
+    """
+    @wraps(method)
+    def _impl(self, *method_args, **method_kwargs):
+        # is there specific name_scope?
+        if hasattr(self, "name"):
+            distr_name = self.name
+        else:
+            distr_name = self.__class__.__name__
+        # is not already scoped with current class to prevent nested scopes due to nested methods
+        if distr_name not in tf.no_op(name='.').name[:-1]:
+            with tf.name_scope(distr_name + "_" + method.__name__):
+                return method(self, *method_args, **method_kwargs)
+        else:
+            return method(self, *method_args, **method_kwargs)
+    return _impl
+
+
+class Distribution(object):
+    """Represent a distribution.
+    """
+
+    def __init__(self, name):
+        self.name = name
+
+    def loglikelihood(self, x, theta):
+        return self._loglikelihood(x, theta)
+
+    @class_scope
+    def mutual_information(self, x, theta):
+        expected_log_qc = self.entropy(x)
+        expected_log_qc_given_x = self.cross_entropy(x, theta)
+        return tf.add(expected_log_qc, expected_log_qc_given_x, name="mutual_information")
+
+    @class_scope
+    def entropy(self, x):
+        return tf.reduce_mean(self.loglikelihood_prior(x), name="entropy")
```

Comment:
Regardless of what is actually used for training, this is a "distribution" class so it should follow strictly the mathematical definition. Entropy should have a negative sign.